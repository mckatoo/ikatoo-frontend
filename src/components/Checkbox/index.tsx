import { InputHTMLAttributes, useState } from 'react'

import Styles from './styles'

export type CheckboxProps = {
  onCheck?: (status: boolean) => void
  isChecked?: boolean
  label?: string
  labelFor?: string
  labelColor?: 'white' | 'black'
  value?: string | ReadonlyArray<string> | number
} & InputHTMLAttributes<HTMLInputElement>

const Checkbox = ({
  onCheck,
  isChecked = false,
  label,
  labelFor = '',
  labelColor = 'white',
  value,
  ...props
}: CheckboxProps) => {
  // controlled component (state)
  const [checked, setChecked] = useState(isChecked)

  const onChange = () => {
    const status = !checked // true => false => true
    setChecked(status)

    !!onCheck && onCheck(status)
  }

  return (
    <Styles.Wrapper>
      <Styles.Input
        id={labelFor}
        type="checkbox"
        onChange={onChange}
        checked={checked}
        value={value}
        {...props}
      />
      {!!label && (
        <Styles.Label htmlFor={labelFor} labelColor={labelColor}>
          {label}
        </Styles.Label>
      )}
    </Styles.Wrapper>
  )
}

export default Checkbox
