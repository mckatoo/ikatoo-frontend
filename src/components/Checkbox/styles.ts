import styled, { css } from 'styled-components'
import tw from 'tailwind-styled-components'
import { CheckboxProps } from '.'

const Input = tw.input`
  border-mck_gray_dark
  delay-100
  ease-in-out

  before:border-white
  before:delay-100
  before:ease-in-out

  focus:shadow-mck_primary

  hover:delay-100
  hover:ease-in-out
  hover:border-gray-500

  checked:border-mck_primary
  checked:bg-mck_primary
  checked:before:opacity-100
`

const Label = tw.label`
  pl-3
`

export default {
  Wrapper: styled.div`
    display: flex;
    align-items: center;
  `,
  Input: styled(Input)`
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid;
    border-radius: 0.2rem;
    transition: background border;
    position: relative;
    outline: none;

    &:before {
      content: '';
      width: 0.6rem;
      height: 0.9rem;
      border: 0.2rem solid;
      border-top: 0;
      border-left: 0;
      transform: rotate(45deg);
      position: absolute;
      top: 0.1rem;
      opacity: 0;
    }

    &:focus {
      box-shadow: 0 0 0.5rem;
    }
  `,
  Label: styled(Label)<Pick<CheckboxProps, 'labelColor'>>`
    ${({ labelColor }) => css`
      cursor: pointer;
      color: ${labelColor};
      line-height: 1.8rem;
    `}
  `
}
