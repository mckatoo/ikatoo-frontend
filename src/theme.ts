export default {
  grid: {
    container: '65rem',
    gutter: '1.6rem'
  },
  border: {
    radius: '0.4rem'
  },
  font: {
    family: "'Oswald', 'Helvetica', 'Arial', 'sans-serif'",
    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xsmall: '0.6rem',
      small: '0.7rem',
      medium: '0.8rem',
      large: '0.9rem',
      xlarge: '1.0rem',
      xxlarge: '1.4rem',
      huge: '2.8rem'
    }
  },
  colors: {
    primary: '#1e293b',
    secondary: '#233d5a',
    mainBg: '#00032c',
    lightBg: '#F2F2F2',
    white: '#FAFAFA',
    black: '#00021f',
    lightGray: '#e5e7eb',
    gray: '#8F8F8F',
    darkGray: '#1f2937',
    red: '#ff2a2a'
  },
  spacings: {
    xxsmall: '0.2rem',
    xsmall: '0.4rem',
    small: '0.6rem',
    medium: '0.8rem',
    large: '1.0rem',
    xlarge: '1.2rem',
    xxlarge: '1.4rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  }
} as const
