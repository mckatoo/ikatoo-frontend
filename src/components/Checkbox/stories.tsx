import { ComponentMeta, ComponentStory } from '@storybook/react'

import Checkbox from '.'

export default {
  title: 'Components/Checkbox',
  component: Checkbox
} as ComponentMeta<typeof Checkbox>

export const Default: ComponentStory<typeof Checkbox> = (args) => (
  <>
    <div style={{ padding: 10 }}>
      <Checkbox
        name="category"
        label="Action"
        labelFor="action"
        isChecked
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <Checkbox
        name="category"
        label="Adventure"
        labelFor="adventure"
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <Checkbox
        name="category"
        label="Strategy"
        labelFor="strategy"
        {...args}
      />
    </div>
  </>
)
