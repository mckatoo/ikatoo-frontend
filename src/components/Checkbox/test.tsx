import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { vi } from 'vitest'

import Checkbox from '.'
import theme from '../../theme'

describe('<Checkbox />', () => {
  it('should render with label', () => {
    const { container } = render(
      <Checkbox label="checkbox label" labelFor="check" />
    )

    expect(screen.getByRole('checkbox')).toBeInTheDocument()
    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument()
    expect(screen.getByText(/checkbox label/i)).toHaveAttribute('for', 'check')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without label', () => {
    render(<Checkbox />)

    expect(screen.queryByLabelText('Checkbox')).not.toBeInTheDocument()
  })

  it('should render with black label', () => {
    render(
      <Checkbox label="checkbox label" labelFor="check" labelColor="black" />
    )

    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: theme.colors.black
    })
  })

  it('should dispatch onCheck when status changes', async () => {
    const onCheck = vi.fn()

    render(<Checkbox label="Checkbox" onCheck={onCheck} />)

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByRole('checkbox'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith(true)
  })

  it('should call onCheck with false if the Checkbox is already checked', async () => {
    const onCheck = vi.fn()

    render(<Checkbox label="Checkbox" onCheck={onCheck} isChecked />)

    userEvent.click(screen.getByRole('checkbox'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith(false)
  })

  it('should be accessible with tab', () => {
    render(<Checkbox label="Checkbox" labelFor="Checkbox" />)

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(screen.getByLabelText(/checkbox/i)).toHaveFocus()
  })
})
