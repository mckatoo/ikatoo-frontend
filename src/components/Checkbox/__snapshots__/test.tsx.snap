// Vitest Snapshot v1

exports[`<Checkbox /> > should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid;
  border-radius: 0.2rem;
  -webkit-transition: background border;
  transition: background border;
  position: relative;
  outline: none;
}

.c1:before {
  content: '';
  width: 0.6rem;
  height: 0.9rem;
  border: 0.2rem solid;
  border-top: 0;
  border-left: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  position: absolute;
  top: 0.1rem;
  opacity: 0;
}

.c1:focus {
  box-shadow: 0 0 0.5rem;
}

.c2 {
  cursor: pointer;
  color: white;
  line-height: 1.8rem;
}

<div
  class="c0"
>
  <input
    class="border-mck_gray_dark delay-100 ease-in-out before:border-white before:delay-100 before:ease-in-out focus:shadow-mck_primary hover:delay-100 hover:ease-in-out hover:border-gray-500 checked:border-mck_primary checked:bg-mck_primary checked:before:opacity-100 c1"
    id="check"
    type="checkbox"
    value=""
  />
  <label
    class="pl-3 c2"
    for="check"
    labelcolor="white"
  >
    checkbox label
  </label>
</div>
`
